### PR TITLE
CI: Use scientific-python/upload-nightly-action

### PR DESCRIPTION
**Pull request by matthewfeickert**
## PR summary

Follow up PR to PR #25950.

* Use the [`scientific-python/upload-nightly-action` GitHub Action](https://github.com/scientific-python/upload-nightly-action) to upload the nightly wheels to the [`scientific-python-nightly-wheels` Anaconda Cloud package index](https://anaconda.org/scientific-python-nightly-wheels). For security best practices pin at the commit sha corresponding to the last stable release and let Dependabot update the commit sha and comment as new releases come out.

* Remove the workflow step for querying and removing all but the last 5 nightly wheel uploads to the scientific-python-nightly-wheels Anaconda Cloud package index as this is now centrally done by the Scientific Python org for all projects that upload to https://anaconda.org/scientific-python-nightly-wheels.
   - c.f. https://github.com/scientific-python/upload-nightly-action/pull/12

## PR checklist

- [N/A] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [N/A] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [N/A] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-examples-and-tutorials)
- [N/A] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [N/A] Documentation complies with [general](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-docstrings) guidelines

### CODE REVIEW

**Comment by matthewfeickert** [comment]
@ksunden @QuLogic This is ready for review. Let me know if you have any questions. :+1:

File: .github/workflows/nightlies.yml [hunk 1]
```diff
@@ -58,38 +58,8 @@ jobs:
...
-              sed 's/.* + //' | \
-              head --lines "-${N_LATEST_UPLOADS}" > remove-package-versions.txt
-
-          while LANG=C IFS= read -r package_version ; do
-              echo "Removing scientific-python-nightly-wheels/matplotlib/${package_version}"
-              anaconda --token ${{ secrets.ANACONDA_ORG_UPLOAD_TOKEN }} remove \
-                --force \
-                "scientific-python-nightly-wheels/matplotlib/${package_version}"
-          done <remove-package-versions.txt
+        uses: scientific-python/upload-nightly-action@8f0394fd2aa0c85d7364a9958652e8994e06b23c # 0.1.0
```
Comment (QuLogic) [inline .github/workflows/nightlies.yml:62]: Is this not tagged? Why not use the tag instead of putting it in a comment?

Comment (matthewfeickert) [reply .github/workflows/nightlies.yml:62]: @QuLogic It is tagged and that can be done if you'd prefer. c.f. https://github.com/scientific-python/upload-nightly-action/issues/7#issuecomment-1564846744 and @henryiii's comment for the motivation for this as it is now.

Comment (ksunden) [reply .github/workflows/nightlies.yml:62]: I think the general principle is that tags can be made to point to a different location, but SHAs cannot (`git checkout` will warn and go to the commit (not the tag)... and fail if the commit does not exist)

So, under the threat model of the tagging authority of upstream being compromised, it is safer to use the SHA.

Comment (matthewfeickert) [reply .github/workflows/nightlies.yml:62]: Yes, thanks for the concise summary @ksunden. Dependabot also knows how to handle this and also update the comment, so as `matplotlib` already has

https://github.com/matplotlib/matplotlib/blob/bfaa6eb677b9c56cafb6a99d6897c9d0cd9d4210/.github/dependabot.yml#L1-L6

you can let Dependabot take care of everything associated to updating the commit SHA.

**Comment by ksunden** [comment]
As far as the removals go, upstream does not yet _actually_ implement deleting, but I'm fine with saying "oh, they will, and if it becomes an actual problem before they do, easy enough to put back, so cross that bridge when we come to it"

File: .github/workflows/nightlies.yml (hunk 1 above)
Comment (matthewfeickert) [reply .github/workflows/nightlies.yml:62]: To be clear, I don't do this for all GitHub Actions, but when it comes to uploading distributions it seems like this is a good time to be overly cautious. I am willing to admit that I also think it is just good to see these patterns exist in more places.

**Comment by matthewfeickert** [comment]
> As far as the removals go, upstream does not yet _actually_ implement deleting, but I'm fine with saying "oh, they will, and if it becomes an actual problem before they do, easy enough to put back, so cross that bridge when we come to it"

Yeah, but as https://github.com/scientific-python/upload-nightly-action/pull/12 will probably be merged in the next 24 hours I don't see any harm in there potentially being 6 wheels up on the package index for a few hours in the worst case scenario. As soon as it is merged I'm also going to trigger a manual run with workflow dispatch.

File: .github/workflows/nightlies.yml (hunk 1 above)
Comment (QuLogic) [reply .github/workflows/nightlies.yml:62]: We should perhaps change `cibuildwheel.yml` to do the same then, because those are what we actually publish...

Comment (matthewfeickert) [reply .github/workflows/nightlies.yml:62]: Opened PR #26025 for this.

Comment (matthewfeickert) [reply .github/workflows/nightlies.yml:62]: Thanks for the fast review and merge on PR #26025 @QuLogic!

**Review by QuLogic — APPROVED** [review]
Seems good to me, but I guess we should wait until at least one nightly wheel is uploaded using the new location, instead of stacking these two changes together.

**Comment by matthewfeickert** [comment]
> but I guess we should wait until at least one nightly wheel is uploaded using the new location, instead of stacking these two changes together.

@ksunden has you covered and did this a few days ago. :) 

https://anaconda.org/scientific-python-nightly-wheels/matplotlib/files

new-package-index

```console
$ docker run --rm -ti python:3.11 /bin/bash
root@e22e7dd70b46:/# python -m pip download --index-url https://pypi.anaconda.org/scientific-python-nightly-wheels/simple --no-deps matplotlib
Looking in indexes: https://pypi.anaconda.org/scientific-python-nightly-wheels/simple
Collecting matplotlib
  Downloading https://pypi.anaconda.org/scientific-python-nightly-wheels/simple/matplotlib/3.8.0.dev1157%2Bg2f778fda79/matplotlib-3.8.0.dev1157%2Bg2f778fda79-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (11.6 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.6/11.6 MB 11.3 MB/s eta 0:00:00
Saved /matplotlib-3.8.0.dev1157+g2f778fda79-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
Successfully downloaded matplotlib
WARNING: There was an error checking the latest version of pip.
root@e22e7dd70b46:/# ls matplotlib-3.8.0.dev1157+g2f778fda79-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl 
matplotlib-3.8.0.dev1157+g2f778fda79-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
root@e22e7dd70b46:/# 
```

**Comment by QuLogic** [comment]
That was manual though, and won't confirm the workflow is working.

**Comment by matthewfeickert** [comment]
> That was manual though, and won't confirm the workflow is working.

Gotcha. :+1: Okay that will happen later tonight then now that PR #25950 is in.

After that happens, if you'd like to see an additional example of the workflow `networkx` has already uploaded [using the same GitHub Action](https://github.com/networkx/networkx/blob/2d9fa51b2baddbfebbf971f366778331309a19d7/.github/workflows/nightly.yml#L20-L23):

[upload](https://github.com/networkx/networkx/actions/runs/5138898711/jobs/9248732202)


[networkx](https://anaconda.org/scientific-python-nightly-wheels/networkx/files)

**Comment by matthewfeickert** [comment]
Oh sorry but you want to check that the API token is correct. Sorry, slow on the uptake today!

**Comment by matthewfeickert** [comment]
And with https://github.com/matplotlib/matplotlib/actions/runs/5151045894/jobs/9275848789 that's up and deploying correctly

```
...
Using Anaconda API: https://api.anaconda.org
Using "scientific-python-nightly-wheels" as upload username
Processing 'dist/matplotlib-3.8.0.dev1192+gbb335a115c-pp39-pypy39_pp73-macosx_10_12_x86_64.whl'
Detecting file type...
File type is "Standard Python"
Extracting standard python attributes for upload
Creating package "matplotlib"
Creating release "3.8.0.dev1192+gbb335a115c"
Uploading file "scientific-python-nightly-wheels/matplotlib/3.8.0.dev1192+gbb335a115c/matplotlib-3.8.0.dev1192+gbb335a115c-pp39-pypy39_pp73-macosx_10_12_x86_64.whl"

 uploaded 0 of 7335Kb: 0.00% ETA: 0.0 minutes
...
 uploaded 8279 of 8279Kb: 100.00% ETA: 0.0 minutes
Upload complete

standard python located at:
https://anaconda.org/scientific-python-nightly-wheels/matplotlib
...
```

```console
$ docker run --rm -ti python:3.11 /bin/bash
root@b6fb237233dd:/# python -m pip download --index-url https://pypi.anaconda.org/scientific-python-nightly-wheels/simple --no-deps matplotlib
Looking in indexes: https://pypi.anaconda.org/scientific-python-nightly-wheels/simple
Collecting matplotlib
  Downloading https://pypi.anaconda.org/scientific-python-nightly-wheels/simple/matplotlib/3.8.0.dev1192%2Bgbb335a115c/matplotlib-3.8.0.dev1192%2Bgbb335a115c-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (11.6 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.6/11.6 MB 6.9 MB/s eta 0:00:00
Saved /matplotlib-3.8.0.dev1192+gbb335a115c-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
Successfully downloaded matplotlib
WARNING: There was an error checking the latest version of pip.
root@b6fb237233dd:/# 
```

:+1:

**Comment by matthewfeickert** [comment]
> Yeah, and as [scientific-python/upload-nightly-action#12](https://github.com/scientific-python/upload-nightly-action/pull/12) will probably be merged in the next 24 hours I don't see any harm in there potentially being 6 wheels up on the package index for a few hours in the worst case scenario. As soon as it is merged I'm also going to trigger a manual run of the wheel removal workflow with workflow dispatch.

https://github.com/scientific-python/upload-nightly-action/pull/12 is now merged and has been run successfully (c.f. https://github.com/scientific-python/upload-nightly-action/pull/12#issuecomment-1584871834).

**Pull request merged**